numpy/core/src/multiarray/dtypemeta.h

```diff
@@ -58,7 +61,13 @@ typedef struct {
      * The casting implementation (ArrayMethod) to convert between two
      * instances of this DType, stored explicitly for fast access:
      */
-    PyObject *within_dtype_castingimpl;
+    PyArrayMethodObject *within_dtype_castingimpl;
+    /*
+     * Implementation which clears a dtype or NULL.  If not given, setting
+     * HASREF on the dtype is invalid.  We only use the loop getting, not
```

Comment:
As far as I can see, `HASREF` isn't a thing, so this comment might be a bit confusing. It looks like `PyDataType_REFCHK` checks `NPY_ITEM_REFCOUNT`, so you could use that? Although maybe `NPY_ITEM_IS_POINTER` is invalid too?

Reply:
Yeah, `NPY_ITEM_REFCOUNT` is it, except that we may want to rename/alias it to `NPY_ITEM_HASREF` or something...

To be honest, I am not sure that `NPY_ITEM_IS_POINTER` is ever used (or useful) except  as having the identical meaning to `NPY_ITEM_REFCOUNT`...

Reply:
Ah that makes sense. I brought up `NPY_ITEM_IS_POINTER` because it might be nice to indicate that a dtype uses embedded references but doesn't need the GIL.